src/subgraph.js

```diff
@@ -17,42 +31,195 @@ module.exports = class Subgraph {
         )
       )
     }
   }
 
+  static validateAbis(manifest, { resolveFile }) {
+    // Validate that the the "source > abi" reference of all data sources
+    // points to an existing ABI in the data source ABIs
+    let abiReferenceErrors = manifest
+      .get('dataSources')
+      .filter(dataSource => dataSource.get('kind') === 'ethereum/contract')
+      .reduce((errors, dataSource, dataSourceIndex) => {
+        let abiName = dataSource.getIn(['source', 'abi'])
+        let abiNames = dataSource.getIn(['mapping', 'abis']).map(abi => abi.get('name'))
+
+        if (abiNames.includes(abiName)) {
+          return errors
+        } else {
+          return errors.push(
+            immutable.fromJS({
+              path: ['dataSources', dataSourceIndex, 'source', 'abi'],
+              message:
+                `ABI name '${abiName}' not found in mapping > abis: ` +
+                `${abiNames.toJS()}`,
+            })
+          )
+        }
+      }, immutable.List())
+
+    // Validate that all ABI files are valid
+    let abiFileErrors = manifest
+      .get('dataSources')
+      .filter(dataSource => dataSource.get('kind') === 'ethereum/contract')
+      .reduce(
+        (errors, dataSource, dataSourceIndex) =>
+          dataSource.getIn(['mapping', 'abis']).reduce((errors, abi, abiIndex) => {
+            try {
+              ABI.load(abi.get('name'), resolveFile(abi.get('file')))
+              return errors
+            } catch (e) {
+              return errors.push(
+                immutable.fromJS({
+                  path: [
+                    'dataSources',
+                    dataSourceIndex,
+                    'mapping',
+                    'abis',
+                    abiIndex,
+                    'file',
+                  ],
+                  message: e.message,
+                })
+              )
+            }
+          }, errors),
+        immutable.List()
+      )
+
+    return abiReferenceErrors.concat(abiFileErrors)
+  }
+
+  static validateContractAddresses(manifest) {
+    return manifest
+      .get('dataSources')
+      .filter(dataSource => dataSource.get('kind') === 'ethereum/contract')
+      .reduce((errors, dataSource, dataSourceIndex) => {
+        let path = ['dataSources', dataSourceIndex, 'source', 'address']
+        let address = dataSource.getIn(['source', 'address'])
+
+        // Validate whether the address has the right length
+        if (
+          !((address.startsWith('0x') && address.length == 42) || address.length == 40)
+        ) {
+          return errors.push(
+            immutable.fromJS({
+              path,
+              message:
+                `Contract address must have length 40 (or 42 if prefixed with 0x) ` +
+                `but has length ${address.length}: ${address}`,
+            })
+          )
+        }
+
+        // Validate whether the address is a hex string
+        let pattern = /^(0x[0-9a-fA-F]{40}|[0-9a-fA-F]{40})$/
```

Comment:
I'm not great with regex, but maybe this be better written as `(0x)?[0-9a-fA-F]{40}`.

This check is a more precise version of the one above it, we could remove the first one, and adjust the error message to mention the accepted formats.

Reply:
Haha, you are right about the regular expression. 👍 Let's combine the two error messages into one like you're suggesting.